server/neptune/workflows/internal/deploy/revision/queue/pull_rebaser.go

```diff
@@ -0,0 +1,159 @@
+	GithubListOpenPRs(ctx context.Context, request activities.ListOpenPRsRequest) (activities.ListOpenPRsResponse, error)
+	GithubListModifiedFiles(ctx context.Context, request activities.ListModifiedFilesRequest) (activities.ListModifiedFilesResponse, error)
+}
+
+type prRevisionSetterActivities interface {
+	SetPRRevision(ctx context.Context, request activities.SetPRRevisionRequest) (activities.SetPRRevisionResponse, error)
+}
+
+type PullRebaser struct {
+	GithubActivities           githubRebaseActivities
+	PrRevisionSetterActivities prRevisionSetterActivities
+}
+
+func (p *PullRebaser) RebaseOpenPRsForRoot(ctx workflow.Context, repo github.Repo, root terraform.Root) error {
+	originalOpts := workflow.GetActivityOptions(ctx)
+
+	// Configure github specific ScheduleToClose timeout for GH API call failures to autoresolve
+	// TODO: Add alarms on GithubListOpenPRs and GithubListModifiedFiles API call failures
+	ghOpts := originalOpts
+	ghOpts.ScheduleToCloseTimeout = RebaseGithubScheduleToCloseTimeout
+	ctx = workflow.WithActivityOptions(ctx, ghOpts)
+
+	// list open PRs
+	var listOpenPRsResp activities.ListOpenPRsResponse
+	err := workflow.ExecuteActivity(ctx, p.GithubActivities.GithubListOpenPRs, activities.ListOpenPRsRequest{
+		Repo: repo,
+	}).Get(ctx, &listOpenPRsResp)
+	if err != nil {
+		return errors.Wrap(err, "listing open PRs")
+	}
+
+	// spawn activities to list modified fles in each open PR in parallel
+	futureByPullNum := map[github.PullRequest]workflow.Future{}
+	for _, pullRequest := range listOpenPRsResp.PullRequests {
+		future := workflow.ExecuteActivity(ctx, p.GithubActivities.GithubListModifiedFiles, activities.ListModifiedFilesRequest{
+			Repo:        repo,
+			PullRequest: pullRequest,
+		})
+		futureByPullNum[pullRequest] = future
+	}
+
+	// resolve the futures and check if an open PR needs to be rebased
+	prsToRebase := []github.PullRequest{}
+	for pullRequest, future := range futureByPullNum {
+		var result activities.ListModifiedFilesResponse
+
+		// list modified files should not fail unless we hit the ratelimit which should autoresolve once our ratelimit revives in an hour max
+		// If it errors out due to any other reason, let's be preventive and rebase this PR as well
+		listFilesErr := future.Get(ctx, &result)
+		if listFilesErr != nil {
+			logger.Error(ctx, "error listing modified files in PR", key.ErrKey, listFilesErr, "pull_num", pullRequest.Number)
+			prsToRebase = append(prsToRebase, pullRequest)
+			continue
+		}
+
+		shouldRebase, err := shouldRebasePullRequest(root, result.FilePaths)
+
+		// unlikey for error since we validate the WhenModified config at startup
+		// if it errors out, let's be preventive and rebase this PR as well
+		if err != nil {
+			logger.Error(ctx, "error matching filepaths in PR", key.ErrKey, err, "pull_num", pullRequest.Number)
+			prsToRebase = append(prsToRebase, pullRequest)
+			continue
+		}
+
+		if shouldRebase {
+			prsToRebase = append(prsToRebase, pullRequest)
```

Comment:
I don't think you should append here. I think instead you should just make the call to rebase within the loop itself.  It's better for us to have rate limit issues after we've rebased X PRs, than to have them after we've rebased 0 PRs.